test/test_common/logging.h

```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "common/common/logger.h"
+
+#include "absl/strings/str_join.h"
+#include "spdlog/spdlog.h"
+
+namespace Envoy {
+
+/**
+ * Provides a mechanism to temporarily set the logging level on
+ * construction, restoring its previous state on dstruction.
+ *
+ * The log_level is the minimum log severity required to print messages.
+ * Messages below this loglevel will be suppressed.
+ *
+ * Note that during the scope of this object, command-line overrides, eg
+ * --log-level trace, will not take effect.
+ *
+ * Also note: instantiating this setter should only occur when the system is
+ * in a quiescent state, e.g. at startup or between tests.
+ *
+ * This is intended for use in EXPECT_LOG_CONTAINS and similar macros.
+ */
+class LogLevelSetter {
+public:
+  explicit LogLevelSetter(spdlog::level::level_enum log_level);
+  ~LogLevelSetter();
+
+private:
+  std::vector<spdlog::level::level_enum> previous_levels_;
+};
+
+/**
+ * Records log messages in a vector<string>, forwarding them to the previous
+ * delegate. This is useful for unit-testing log messages while still being able
+ * to see them on stderr.
+ *
+ * Also note: instantiating this sink should only occur when the system is
+ * in a quiescent state, e.g. at startup or between tests
+ *
+ * This is intended for use in EXPECT_LOG_CONTAINS and similar macros.
+ */
+class LogRecordingSink : public Logger::SinkDelegate {
+public:
+  explicit LogRecordingSink(Logger::DelegatingLogSinkPtr log_sink);
+  virtual ~LogRecordingSink();
+
+  // Logger::SinkDelgate
+  void log(absl::string_view msg) override;
+  void flush() override;
+
+  const std::vector<std::string>& messages() const { return messages_; }
+
+private:
+  std::vector<std::string> messages_;
+};
+
+// Validates that when stmt is executed, exactly one log message containing substr will be emitted.
+#define EXPECT_LOG_CONTAINS(substr, stmt)                                                          \
```

Comment:
Might we want to specify what log level we're expecting the string to appear at as well, e.g. if the documented behavior is we log at "warn", then we should test at "warn" rather than trace.

Reply:
OK, ptal.  Not sure if this is the best way to do this, given I've hooked in after the string has been constructed.